packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,204 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+pragma experimental ABIEncoderV2;
+
+import "../../iface/IExchangeV3.sol";
+
+import "../../lib/Claimable.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/ReentrancyGuard.sol";
+import "../../lib/SignatureUtil.sol";
+import "../../lib/EIP712.sol";
+
+
+/// @title Fast withdrawal agent implementation.
+///
+///        A mediator/liquidity provider (whoever provides the funds immediately) can call
+///        `executeFastWithdrawals` to provide users immediately with funds onchain.
+///        This allows the security of the funds to be handled by any EOA or smart contract.
+///
+///        Users that want to make use of this functionality have to
+///        authorize this contract as their agent.
+///
+///        TODO: generalize this (here or another agent) with any function call
+///        a user wants to do to be able to simulate completely onchain smart wallet with
+///        funds stored in the Loopring protocol.
+///
+/// @author Brecht Devos - <brecht@loopring.org>
+contract FastWithdrawalAgent is ReentrancyGuard
+{
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using SignatureUtil     for bytes32;
+
+    struct FastWithdrawal
+    {
+        address exchange;
+        address from;                   // The owner of the account
+        address to;                     // The address that will receive the tokens withdrawn
+        address token;
+        uint    amount;
+        address feeToken;
+        uint    fee;
+        bool    onchainFeePayment;
+        uint    validUntil;
+        bytes   signature;
+    }
+
+    // Replay protection
+    mapping (bytes32 => bool) txHash;
+
```

Comment:
It will be great if we can use onchain withdrawal request for fast withdrawal. So the request can express:

_I want to withdrawal x token Y, if someone can make it fast within x minutes (this should be configured by the exchange not customized by users), I'll pay extra z% fee._

So if nobody can facility the fast withdrawal, it falls back to normal withdrawal.

Reply:
This is a good idea. I added this to the agent. I think the main benefit is that the exchange wouldn't need to bother the user again with a new request to sign if it wasn't able to get the fast withdrawal through in time.